### PR TITLE
Hunger / Exhaustion / Food properties re-implemented.

### PR DESCRIPTION
**Pull request by greatman**
Signed-off-by: William Lebel lebelwilliam@gmail.com

### CODE REVIEW

File: src/main/java/org/spout/vanilla/components/living/Human.java [hunk 1]
```diff
@@ -139,18 +141,7 @@ public void setGamemode(GameMode mode) {
...
-			case CREATIVE:
-				holder.add(CreativeComponent.class);
-				break;
-			case SURVIVAL:
-				holder.add(SurvivalComponent.class);
-				break;
-		}
-		holder.getNetwork().callProtocolEvent(new PlayerGameStateEvent( (Player) holder, PlayerGameStateMessage.CHANGE_GAME_MODE, mode));
+		holder.getNetwork().callProtocolEvent(new PlayerGameStateEvent( (Player) holder, PlayerGameStateMessage.CHANGE_GAME_MODE, event.getMode()));
 		getData().put(VanillaData.GAMEMODE, mode);
```
Comment (Darsstar) [inline src/main/java/org/spout/vanilla/components/living/Human.java]: mode should be event.getMode()

**Comment by greatman** [comment]
Updated it for latest stuff. I still got 2 bugs with it:
- The item amount is not changed after eating. Can't seem to find why tho :/
- Health is stuck at 1 due to MaxHealth being wrong.

**Comment by windy1** [comment]
Hm, It's possible that window's are not being updated properly. I will look into that for you.

**Comment by greatman** [comment]
@W1ndwaker Thanks a ton.

**Comment by greatman** [comment]
And finally done. Everything works!

Also removed some unused imports in the code.

File: src/main/java/org/spout/vanilla/component/living/Human.java [hunk 2]
```diff
@@ -142,20 +143,7 @@ public void setGamemode(GameMode mode) {
 				return;
 			}
 			mode = event.getMode();
-		}
```
Comment (windy1) [inline src/main/java/org/spout/vanilla/component/living/Human.java]: Why did you remove this?

Comment (greatman) [inline src/main/java/org/spout/vanilla/component/living/Human.java]: The switch? Zidane asked to scrap the components for gamemode some weeks ago.

**Comment by greatman** [comment]
Updated with latest changes.

File: src/main/java/org/spout/vanilla/component/living/Human.java [hunk 3]
```diff
@@ -59,9 +57,15 @@ public void onAttached() {
 		Entity holder = getHolder();
 		holder.add(PickupItemComponent.class);
 		holder.add(DiggingComponent.class);
+
+		if (getHolder() instanceof Player) {
+			holder.add(HungerComponent.class);
```
Comment (windy1) [inline src/main/java/org/spout/vanilla/component/living/Human.java]: we want to set this in the Listener please

File: src/main/java/org/spout/vanilla/component/player/HungerComponent.java [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+			}
+		} else {
+			counter++;
+		}
+	}
+
+	/**
+	 * Update the player stats client-side.
+	 */
+	public void updatePlayer() {
```
Comment (windy1) [inline src/main/java/org/spout/vanilla/component/player/HungerComponent.java]: We probably want this to be automatic, no? We probably don't even need a function for it.

Comment (windy1) [inline src/main/java/org/spout/vanilla/component/player/HungerComponent.java]: Ah, I see, never mind.

**Comment by windy1** [comment]
Everything tested, and is the eating animation working?

**Comment by greatman** [comment]
Didin't check for the animation. Will do tomorrow.

File: src/main/java/org/spout/vanilla/component/living/Human.java (hunk 3 above)
Comment (greatman) [inline src/main/java/org/spout/vanilla/component/living/Human.java]: Listener?

Comment (windy1) [inline src/main/java/org/spout/vanilla/component/living/Human.java]: Yes, instead of adding the component in human, add it onPlayerJoin in VanillaListener please. :)

Comment (greatman) [inline src/main/java/org/spout/vanilla/component/living/Human.java]: Oh I see. Will do ;)

File: src/main/java/org/spout/vanilla/component/player/HungerComponent.java [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+	@Override
+	public boolean canTick() {
+		return getHolder().get(Human.class).isSurvival();
+	}
+
+	@Override
+	public void onTick(float dt) {
+
+		float distanceMoved = 0f;
+		Point position = getHolder().getTransform().getPosition();
```
Comment (Zidane) [inline src/main/java/org/spout/vanilla/component/player/HungerComponent.java]: TransformComponent has the following methods...

getTransform (old snapshot value)
getTransformLive (live value that is current of this tick)
isPositionDirty (if live != old)

You don't need to store position yourself.

File: src/main/java/org/spout/vanilla/VanillaListener.java [hunk 6]
```diff
@@ -63,5 +64,6 @@ public void onPlayerJoin(PlayerJoinEvent event) {
 		player.add(PlayerListComponent.class);
 		player.add(PingComponent.class);
 		player.add(PickupItemComponent.class);
+		player.add(HungerComponent.class);
 	}
```
Comment (Zidane) [inline src/main/java/org/spout/vanilla/VanillaListener.java]: Entities should be able to have the hunger concept.

File: src/main/java/org/spout/vanilla/component/player/HungerComponent.java [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+	private Point oldPosition = null;
+
+	@Override
+	public boolean isDetachable() {
+		return false;
+	}
+
+	@Override
+	public boolean canTick() {
+		return getHolder().get(Human.class).isSurvival();
```
Comment (Zidane) [inline src/main/java/org/spout/vanilla/component/player/HungerComponent.java]: Update this for entities and when you do, make a flag in HungerComponent to toggle hunger off (or perhaps setting hunger to -1 acts as a disabler).

Better as a flag (think HealthComponent hasInfiniteHealth but that needs to be made a boolean...).

File: src/main/java/org/spout/vanilla/component/player/HungerComponent.java [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+				if (getFoodSaturation() > 0.0f) {
+					setFoodSaturation(Math.max(getFoodSaturation() - 1f, 0), FoodSaturationChangeCause.EXHAUSTION);
+				} else {
+					setHunger((short) Math.max(getHunger() - 1, 0), HungerChangeCause.LOW_FOOD_SATURATION);
+				}
+			}
+
+			if (getHunger() <= 0 && !getHolder().get(HealthComponent.class).isDead()) {
+				int maxDrop;
+				switch (getHolder().getWorld().getComponentHolder().getData().get(VanillaData.DIFFICULTY)) {
```
Comment (Zidane) [inline src/main/java/org/spout/vanilla/component/player/HungerComponent.java]: Hmm, when this is updated for entities, have difficulty affect entities too? Just Players? Situational(entities damage increases with difficulty but only that)

File: src/main/java/org/spout/vanilla/component/player/HungerComponent.java [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+
+	/**
+	 * Sets the hunger (food bar value) of the player.
+	 * 
+	 * @param hunger The level of hunger to set to.
+	 * @param source The source of the change.
+	 */
+	public void setHunger(short hunger, Source source) {
+		PlayerHungerChangeEvent event = new PlayerHungerChangeEvent((Player) getHolder(), source, (short) (hunger - getHunger()));
+		Spout.getEngine().getEventManager().callEvent(event);
```
Comment (Zidane) [inline src/main/java/org/spout/vanilla/component/player/HungerComponent.java]: 1. Only call this if PlayerHungerChangeEvent.getHandlerList().getRegisteredListeners().length > 0
2. You can condense this down to one line: PlayerHungeChangeEvent event = Spout.get....

**Comment by Zidane** [comment]
Looking back on my comments above, might be even better to extend HungerComponent for a player version

File: src/main/java/org/spout/vanilla/component/living/VanillaEntity.java [hunk 10]
```diff
@@ -41,6 +42,8 @@ public void onAttached() {
 		holder.add(HeadComponent.class);
 		holder.add(HealthComponent.class);
 		holder.add(VanillaPhysicsComponent.class);
+		holder.add(HungerComponent.class);
```
Comment (Zidane) [inline src/main/java/org/spout/vanilla/component/living/VanillaEntity.java]: Two expensive calls...but for price of two?

How about HungerComponent hunger = holder.add(HungerComponent.class);

File: src/main/java/org/spout/vanilla/component/misc/HungerComponent.java [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+				if (getFoodSaturation() > 0.0f) {
+					setFoodSaturation(Math.max(getFoodSaturation() - 1f, 0), FoodSaturationChangeCause.EXHAUSTION);
+				} else {
+					setHunger((short) Math.max(getHunger() - 1, 0), HungerChangeCause.LOW_FOOD_SATURATION);
+				}
+			}
+
+			if (getHunger() <= 0 && !getOwner().get(HealthComponent.class).isDead()) {
+				int maxDrop;
+				switch (getOwner().getWorld().getComponentHolder().getData().get(VanillaData.DIFFICULTY)) {
```
Comment (Zidane) [inline src/main/java/org/spout/vanilla/component/misc/HungerComponent.java:100]: getOwner().getData()

File: src/main/java/org/spout/vanilla/component/misc/HungerComponent.java [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+					break;
+				default:
+					maxDrop = 0;
+				}
+				if (maxDrop < getOwner().get(HealthComponent.class).getHealth()) {
+					getOwner().get(HealthComponent.class).setHealth(getOwner().get(HealthComponent.class).getHealth() - 1, HealthChangeCause.NO_FOOD);
+				}
+
+			} else if (getHunger() >= 18 && getOwner().get(HealthComponent.class).getHealth() < 20) {
+				getOwner().get(HealthComponent.class).setHealth(getOwner().get(HealthComponent.class).getHealth() + 1, HealthChangeCause.REGENERATION);
```
Comment (Zidane) [inline src/main/java/org/spout/vanilla/component/misc/HungerComponent.java]: Two for the price of two :P (see above)

File: src/main/java/org/spout/vanilla/protocol/handler/player/PlayerBlockPlacementHandler.java [hunk 13]
```diff
@@ -213,7 +212,7 @@ public void handleServer(Session session, PlayerBlockPlacementMessage message) {
 					sound.playGlobal(target.getPosition(), 0.8f, 0.8f);
 					//GeneralEffects.BREAKBLOCK.playGlobal(target.getPosition(), target.getMaterial());
 					// Remove block from inventory if not in creative mode.
-					if (!player.has(CreativeComponent.class)) {
+					if (!player.get(Human.class).isCreative()) {
```
Comment (Zidane) [inline src/main/java/org/spout/vanilla/protocol/handler/player/PlayerBlockPlacementHandler.java]: Should be add

File: src/main/java/org/spout/vanilla/protocol/handler/player/PlayerPositionHandler.java [hunk 14]
```diff
@@ -60,5 +63,20 @@ public void handleServer(Session session, PlayerPositionMessage message) {
 				}
 			}
 		}
+		
+		Human human = holder.get(Human.class);
```
Comment (Zidane) [inline src/main/java/org/spout/vanilla/protocol/handler/player/PlayerPositionHandler.java]: Should be add

File: src/main/java/org/spout/vanilla/VanillaListener.java [hunk 15]
```diff
@@ -65,5 +66,6 @@ public void onPlayerJoin(PlayerJoinEvent event) {
 		player.add(PingComponent.class);
 		player.add(PickupItemComponent.class);
 		player.add(SleepComponent.class);
+		player.get(HungerComponent.class).setInfiniteHunger(false);
```
Comment (Zidane) [inline src/main/java/org/spout/vanilla/VanillaListener.java]: should be add

File: src/main/java/org/spout/vanilla/component/misc/HungerComponent.java [hunk 16]
```diff
@@ -0,0 +1,271 @@
...
+	public void onTick(float dt) {
+
+		float distanceMoved = 0f;
+		if (getOwner().getTransform() != null) {
+			distanceMoved += getOwner().getTransform().getTransformLive().getPosition().distanceSquared(getOwner().getTransform().getTransform().getPosition());
+			if (distanceMoved >= 1) {
+				addExhaustion(ExhaustionLevel.WALKING.getAmount());
+			}
+		}
+		if (getOwner().has(Human.class)) {
```
Comment (Zidane) [inline src/main/java/org/spout/vanilla/component/misc/HungerComponent.java]: Instead of has...

Human human = entity.get(Human.class);
if (human != null)

**Comment by Zidane** [comment]
Needs to be updated to HEAD

**Comment by greatman** [comment]
Updated it. Just need to quick test it.

File: src/main/java/org/spout/vanilla/component/living/LivingComponent.java [hunk 17]
```diff
@@ -43,6 +44,8 @@ public void onAttached() {
 		holder.add(HealthComponent.class);
 		holder.add(VanillaPhysicsComponent.class);
 		holder.add(DrowningComponent.class);
+		HungerComponent hunger = holder.add(HungerComponent.class);
```
Comment (windy1) [inline src/main/java/org/spout/vanilla/component/living/LivingComponent.java:47]: Only human's have hunger, no?

Comment (greatman) [inline src/main/java/org/spout/vanilla/component/living/LivingComponent.java:47]: Zidane asked some commits ago to have it for Entities.

**Comment by Afforess** [comment]
This commit is way out of date and will probably need to be rewritten.
